### PR TITLE
Verify image signatures policy

### PR DESCRIPTION
**Pull request by raulcabello**
## Description

Verify all containers (including init and ephemeral) that matches the image name pattern provided. For that it will call the validate functions from the policy sdk.

The policy takes a list of signatures. A signature can be of two types: public key or keyless. Each signature
has an `image` field which will be used to select the matching containers in the pod that will be evaluated.
`image` supports wildcard. For example, `ghcr.io/kubewarden/*` will match all images from the kubewarden ghcr repo.

Fix kubewarden/policy-server#99

### CODE REVIEW

**Comment by raulcabello** [comment]
waiting for [sigstore docs](https://github.com/kubewarden/docs/pull/100)

**Review by viccuad — APPROVED** [review]
LGTM! approving in advance of sdk changes

**Comment by raulcabello** [comment]
you can now review it, but please don't merge until the [sdk](https://github.com/kubewarden/policy-sdk-rust/pull/32) and [sigstore docs](https://github.com/kubewarden/docs/pull/100) are merged

File: README.md
Comment (viccuad) [inline README.md:10]: trailing dot missing :D
```suggestion
This mutation can be disabled by setting `modifyImagesWithDigest` to `false`.
```

File: src/lib.rs
Comment (viccuad) [inline src/lib.rs:118]: Why not return a `Result<PodSpec>` instead? and then use a match in `validate()`

**Review by viccuad — APPROVED** [review]
LGTM :). Just a comment.

File: src/lib.rs
Comment (raulcabello) [reply src/lib.rs]: great idea! done in https://github.com/kubewarden/verify-image-signatures/pull/4/commits/3f8ff915481f74266e13c89fa7c0beb0e65a5ed8

**Review by viccuad — APPROVED** [review]

File: Cargo.toml
Comment (flavio) [inline Cargo.toml:13]: I just learnt the `api` feature causes a longer compilation time and a bigger binary, but we don't actually need that inside of the policies. This introduces all the functions required to get/patch/delete Kubernetes objects

```suggestion
k8s-openapi = { version = "0.11.0", default_features = false, features = ["v1_20"] }
```

File: README.md
Comment (flavio) [inline README.md:5]:
```suggestion
This policy validates Sigstore signatures for containers, init container and ephemeral container that match the name provided
```

File: README.md
Comment (flavio) [inline README.md:6]:
```suggestion
in the `image` settings field. It will reject the Pod if any validation fails.
```

File: README.md
Comment (flavio) [inline README.md:7]:
```suggestion
If all signature validation pass or there is no container that matches the image name, the Pod will be accepted.
```

File: src/lib.rs
Comment (flavio) [inline src/lib.rs:80]: This assumes the users never reference images by digest. This could lead to mutation requests responses that are not changing the Pod. This is not causing harm, but I think it would be better to reply with a `mutate_request` only when the contents of the Pods have been actually changed

File: src/lib.rs
Comment (flavio) [inline src/lib.rs:88]:
```suggestion
                                "Pod {} is not accepted: {}",
```

File: src/lib.rs
Comment (flavio) [inline src/lib.rs:97]: Maybe make it a `debug!` or even drop it entirely

File: src/lib.rs
Comment (flavio) [inline src/lib.rs:110]: Maybe we could change the return signature to be:

```rust
Result<Option<PodSpec>, String>
```
With that:

  * `Ok(None)` -> accept, without any change
  * `Ok(Some(pod_spec))` -> accept, but with mutation

File: Cargo.toml
Comment (flavio) [inline Cargo.toml:22]: This looks like a dev dependency

File: src/lib.rs
Comment (flavio) [inline src/lib.rs:289]: There's a nicer way to declare that, `serde_json` provides a `json!` macro that can be used like that:

```rust
let my_serde_json_value: serde_json::Value = json!({
  "hello": "world",
  "times": 2,
});
```

File: src/lib.rs
Comment (flavio) [inline src/lib.rs:392]: Same here

File: src/settings.rs
Comment (flavio) [inline src/settings.rs:13]: I think we can drop these lines

**Review by flavio — COMMENTED** [review]
Great stuff, I left some comments

File: src/lib.rs
Comment (raulcabello) [reply src/lib.rs:81]: yes, that makes sense. done in https://github.com/kubewarden/verify-image-signatures/pull/4/commits/6cf9839979d4e5e78642006d40e7c373a65f4418

File: src/lib.rs
Comment (raulcabello) [reply src/lib.rs]: done in https://github.com/kubewarden/verify-image-signatures/pull/4/commits/6cf9839979d4e5e78642006d40e7c373a65f4418

**Comment by raulcabello** [comment]
@flavio I made changes with you suggestions, please have a look. Thanks!

**Review by flavio — APPROVED** [review]
LGTM. Thanks for the changes!

**Review by jvanz — APPROVED** [review]

**Pull request merged**
